### ISSUE TITLE
Error setting storage-opt size in btrfs on new container.

### ISSUE THREAD

**Issue by mauri** [moby/moby#33627]
**Description**

Error while passing storage-opt to docker run to limit the size of the container rootfs using btrfs.

**Steps to reproduce the issue:**
1. initialize docker to use the btrfs storage driver as described in https://docs.docker.com/engine/userguide/storagedriver/btrfs-driver/#configure-docker-to-use-the-btrfs-storage-driver
2. docker run -ti --storage-opt size=1G alpine
3. got Error response from daemon: Failed to limit qgroup for &{}: invalid argument.

**Describe the results you received:**

docker: Error response from daemon: Failed to limit qgroup for &{}: invalid argument.
See 'docker run --help'.

**Describe the results you expected:**

A running container whose rootfs is limited to 1G :)

**Additional information you deem important (e.g. issue happens only occasionally):**

This is a consistent issue.

**Output of `docker version`:**

```
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:06 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:06 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 1
Server Version: 17.05.0-ce
Storage Driver: btrfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9048e5e50717ea4497b757314bad98ea3763c145
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 apparmor
Kernel Version: 4.4.0-66-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.959GiB
Name: lubuntu
ID: 2RCA:JFZP:O6BW:Y4OS:SROP:JBXA:BWKU:ENOV:WUSN:QMC7:ZKH7:
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

**Comment by thaJeztah** [comment]
Is the daemon also configured with [`storage-opt: btrfs.min_space`](https://github.com/moby/moby/blob/67fdf574d5acd6ddccb6ece0ffe0ace1c1608712/docs/reference/commandline/dockerd.md#btrfs-options) ? I seem to recall they had to be both set, but perhaps I'm mistaken

ping @yongtang perhaps you know :smile:

(if that option indeed has to be set, we should at least try to make a better error-message)

**Comment by mauri** [comment]
@thaJeztah thanks, tried that but still not working
```
root@ubuntu:/etc/docker# cat /etc/docker/daemon.json
{
  "storage-driver": "btrfs",
  "storage-opt": "btrfs.min_space=1G"
}
root@lubuntu:/etc/docker# service docker-engine restart
root@ubuntu:/etc/docker# docker run -ti --storage-opt size=2G alpine
docker: Error response from daemon: Failed to limit qgroup for &{}: invalid argument.
See 'docker run --help'.
```

**Comment by yongtang** [comment]
@mauri I tried with the most recent master build and it works fine:
```
# In Terminal 1

ubuntu@ubuntu:~/docker$ sudo modprobe btrfs
ubuntu@ubuntu:~/docker$ sudo cat /proc/filesystems | grep btrfs
	btrfs
ubuntu@ubuntu:~/docker$ sudo apt-get install btrfs-tools
Selecting previously unselected package btrfs-progs.
(Reading database ... 110054 files and directories currently installed.)
...
ubuntu@ubuntu:~/docker$ sudo rm -rf /var/lib/docker
ubuntu@ubuntu:~/docker$ sudo mkdir /var/lib/docker
ubuntu@ubuntu:~/docker$ sudo mkfs.btrfs -f /dev/sdb 
btrfs-progs v4.7.3
See http://btrfs.wiki.kernel.org for more information.

Label:              (null)
UUID:               
Node size:          16384
Sector size:        4096
Filesystem size:    20.00GiB
Block group profiles:
  Data:             single            8.00MiB
  Metadata:         DUP               1.00GiB
  System:           DUP               8.00MiB
SSD detected:       no
Incompat features:  extref, skinny-metadata
Number of devices:  1
Devices:
   ID        SIZE  PATH
    1    20.00GiB  /dev/sdb

ubuntu@ubuntu:~/docker$ sudo mount -t btrfs /dev/sdb /var/lib/docker
ubuntu@ubuntu:~/docker$ sudo dockerd --debug --storage-driver btrfs
DEBU[0000] Listener created for HTTP on unix (/var/run/docker.sock) 
INFO[0000] libcontainerd: new containerd process, pid: 13196 
...
```

```
# In Terminal 2

ubuntu@ubuntu:~/docker$ docker run -ti --storage-opt size=1G alpine
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
2aecc7e1714b: Pull complete 
Digest: sha256:0b94d1d1b5eb130dd0253374552445b39470653fb1a1ec2d81490948876e462c
Status: Downloaded newer image for alpine:latest
/ # 

```

```
# In Terminal 3

ubuntu@ubuntu:~/docker$ sudo btrfs qgroup show /var/lib/docker
qgroupid         rfer         excl 
--------         ----         ---- 
0/5         152.00KiB    152.00KiB 
0/257         4.04MiB     96.00KiB 
0/258         4.04MiB     80.00KiB 
0/259         4.04MiB     96.00KiB
```

**Comment by mauri** [comment]
Thanks @yongtang, this seems to work in 17.03.1-ce, and now with master, but not in the 17.05.0-ce release.

**Comment by cpuguy83** [comment]
Yes b36e613d9f311e69387ccec2be16f8618fa1f558 was merged in 17.04 which likely broke it.
Fixed by e907c6418a2ae7754f69fede5897a7176250e8bd and 16328cc207a493ecff0cabc11ebf51e958131418

Closing since this looks resolved.
